### ISSUE TITLE
Support installing separate Argo CD servers for each project

### ISSUE THREAD

**Issue by noam-codefresh** [argoproj-labs/argocd-autopilot#65]
This can be optional (the bootstrap server **may** manage all/some of the projects)
Create a folder structure where each different server has a separate "root" application, with its own list of (one or more) projects.
Identical apps on different projects (which may be managed by different argo-cd servers, and running on different k8s clusters) will have **the same** base

**Comment by joebowbeer** [comment]
Are the different ArgoCD servers all running the same version?

What changes are needed to the bootstrap command in order to support this?

Does this issue cover the use case where several clusters are bootstrapped to different folders of  the same repo?

What folder structure changes are needed to support these use cases?

```
bootstrap/
    <cluster-name>/
        argo-cd/
        argo-cd.yaml
        root.yaml
```

**Comment by noam-codefresh** [comment]
yeah, the directory structure you suggest seems like the direction we were thinking. I think that the `projects` folder might also be moved into the `bootstrap/<cluster-name>/` directory, so at the end the installation-path will only contain two directories - `bootstrap` and `apps`.
We also thought about renaming the `bootstrap` directory to `installations`.

> Does this issue cover the use case where several clusters are bootstrapped to different folders of the same repo?

Do you mean after running `repo bootstrap --installation-path path1`, and then `repo bootstrap --installation-path path2`? After those two commands, there will be completely separate installations in the same repo, and they will not relate to each other. They just happen to be committed to the same repo.
The usage of `--installation-path` is mainly for users who already have an existing GitOps repo, and they don't want the autopilot directories to be at its root.

**Comment by joebowbeer** [comment]
Regarding bootstrapping several clusters to separate folders, I was thinking of building a fleet repo incrementally.

First I would connect the staging cluster to one folder and then I would connect the prod cluster to a sibling folder.

Another dimension to consider is whether ArgoCD is running in each cluster or only in a separate management cluster.

**Comment by renaudguerin** [comment]
This would be super useful, any progress on this ?

**Comment by roi-codefresh** [comment]
Hey @renaudguerin, we just had a quick talk about this feature with the team and we think about separating the CLI into two modes of operation:
1. single argocd - multi destination clusters
2. multi argocd - single destination cluster.

We think that supporting multi-argocd - multi destination clusters might become a mess pretty quickly.

Assuming you choose to go with multi argocd - single destination cluster we would create the directory structure as follows:
```
bootstrap/
    cluster1/
        argo-cd/
        root.yaml   ----------
    cluster2/                |
        argo-cd/             |
        root.yaml   -------| |
                           | |    includes (project1 and project2) - add projects to an argocd installation using 'project add' command
projects/                  | |    single project can be shared for multiple installations.
    project1/       <--------|
       ...                 | |
    project2/       <======|=|
       ...                 |
    project3/       <------|  includes (project2 and project3)
       ...
```

What do you think about this scenario? Did you have this in mind or something different?

**Comment by renaudguerin** [comment]
> 1. single argocd - multi destination clusters
> 2. multi argocd - single destination cluster.
> We think that supporting multi-argocd - multi destination clusters might become a mess pretty quickly.

Yes, that sounds sensible.

> Assuming you choose to go with multi argocd - single destination cluster we would create the directory structure as follows:
> [...]
> What do you think about this scenario? Did you have this in mind or something different?

The bootstrap part looks great for a multi-argocd + single destination setup.
How would you point to individual projects from each cluster's `root.yaml` though ? Currently it's an app of apps pattern (well, app of "appsets + appprojects") pointing to `projects/`. Should it point to a `clusterN` directory under `projects` instead ? I don't understand why your scenario has `projectN` *directories* instead.

When it comes to projects:
I'm still not sure how to use them for multiple dimensions at the same time (user/system apps x dev/production) as I explained in #140 

At the moment (single ArgoCD, single dev environment) my directory structure looks like this : 
```
.
├── apps/
│   ├── cluster-components/
│   │   ├── cert-manager/
│   │   │   ├── config.json
│   │   │   ├── Chart.yaml <- Wrapper Helm chart, workaround for https://github.com/argoproj/argo-cd/issues/2789
│   │   │   └── values.yaml
│   │   ├── nginx/
│   │   │   ├── config.json
│   │   │   ├── Chart.yaml
│   │   │   └── values.yaml
│   │   └── external-secrets/
│   │       ├── config.json
│   │       ├── Chart.yaml
│   │       └── values.yaml
│   └── user-apps/
│       └── example-service/
│           ├── base/
│           └── overlays/
│               └── development/
│                   ├── config.json
│                   └── kustomization.yaml
├── bootstrap/
│   ├── argo-cd/
│   │   ├── argocd-server-ingress.yaml <- ingress (hardcoded hostname)
│   │   ├── cf-argocd-agent.yaml <- generated by "codefresh install gitops argocd-agent"
│   │   └── kustomization.yaml <- manually tweaked to add resources + ArgoCD RBAC accounts
│   ├── cluster-resources/
│   │   ├── in-cluster/
│   │   │   ├── argocd-ns.yaml 
│   │   │   ├── codefresh-role-sa-bind.yaml
│   │   │   ├── config-connector.yaml <- GCP Config Connector setup with env-specific service account
│   │   │   └── development-ns.yaml <- development namespace with annotation for GCP Config Connector
│   │   └── in-cluster.json
│   ├── argo-cd.yaml
│   ├── cluster-resources.yaml
│   └── root.yaml
└── projects/
    ├── cluster-components.yaml <- tweaked with "path: apps/cluster-components/**/config.json" and "source: helm:"
    └── user-apps.yaml
```

As you can see my "system" apps (cluster-components) are all Helm apps (in fact, wrapper Helm charts mostly here to provide values [until ArgoCD provides a better solution](https://github.com/argoproj/argo-cd/issues/2789), plus in the case of `cert-manager` create some resources like Issuers). That's why I've manually removed the reference to overlays in the ApplicationSet `path` in `cluster-components.yaml` (and tweaked the App spec to use a Helm source)

I need to go to production now and I'm not yet sure how to tweak this to support multiple environments : ideally just add a `values-production.yaml` and point `spec.template.spec.source.helm.valuesFiles` in the ApplicationSet to it.
Hopefully I don't even need a separate ApplicationSet/project to do that if something like `- values-{{ destNamespace }}.yaml` can be used for `valueFiles`.
But maybe values won't be enough and I'll need to turn these apps into [Kustomized Helm apps](https://github.com/argoproj/argocd-example-apps/blob/master/plugins/kustomized-helm/README.md), in which case overlays will make sense again and I can remove my tweaks to the autopilot-generated project.

When it comes to the `user-apps` project, I wonder if the same parameterisation trick could be used for environments, i.e. `path: apps/user-apps/**/overlays/{{ destNamespace }}/config.json` for a single project yaml file, instead of env-specific ones ? But I suppose a single project file isn't so valuable if `root.yaml` can easily point to different ones for each cluster.

**Comment by renaudguerin** [comment]
Also, what would be a good migration plan to this proposed new directory structure ? Manually disable `selfHeal` in `autopilot-bootstrap`, commit directory structure change, change path to `bootstrap/clusterN` directory, reenable `selfHeal` ?

**Comment by renaudguerin** [comment]
@roi-codefresh Any thoughts on the above ?

I need to move forward, so I'm going to try out the directory structure you suggested + manually apply tweaked manifests from autopilot dry runs, and see what I can come up with.

**Comment by myspotontheweb** [comment]
In my case I run an ArgoCD instance on each cluster. Wouldn't it be simpler to just install the Project and ApplicationSet objects into the target clusters as follows:

````
# Deploy "dev" to cluster1
kubectl apply --context cluster1 --name argocd -f https://raw.githubusercontent.com/myspotontheweb/gitops-workloads/master/projects/dev.yaml

# Deploy "test" to cluster2
kubectl apply --context cluster2 --name argocd -f https://raw.githubusercontent.com/myspotontheweb/gitops-workloads/master/projects/test.yaml

# Deploy "prod" to cluster3
kubectl apply --context cluster3 --name argocd -f https://raw.githubusercontent.com/myspotontheweb/gitops-workloads/master/projects/prod.yaml
````

**Comment by diegoluisi** [comment]
Hey everyone,
I’m looking to bootstrap ArgoCD in standalone mode using `argocd-autopilot`, but with a **single repository**. I found [this issue](https://github.com/argoproj-labs/argocd-autopilot/issues/65) still open, and I wanted to check if there’s been any progress or workarounds.
My setup:
- Multiple clusters in different environments **without connectivity between them**
- I **don’t want to create multiple repositories**
- Ideally, I’d like to have a **Kustomization per cluster** inside `bootstrap/argocd/overlays/clustername/`
- Then use **AppSets** to deploy workloads for each environment
Is this possible with `argocd-autopilot` today? Or would I need to handle the overlays manually outside of Autopilot?
Thanks in advance!
